rfcs/004 - Continuous Two-Way Synchronization RFC.md

```diff
@@ -0,0 +1,209 @@
+
+These checks should also pay attention to the timestamp and ignore the missing `end` record condition if it has been "too long" and therefore the engine was likely restarted in the meantime. The exact time will depend on what limit we ourselves put on the new commit logic, but perhaps 45 minutes (the current apply/sync timeout time).
+
+It is recommended to keep `iasql_apply`, `iasql_sync`, `iasql_preview_apply`, and `iasql_preview_sync` as functions that can be called, but make them no-ops that warn/error that they don't do anything anymore, and that at some time in the future these functions will be dropped. The reasoning behind this is that this transition will take some time and there may be IaSQL snippets out there that do continue working after this transition assuming they didn't call either of those functions. We can track the number of actual invocations in these functions by having the engine-side implementation simply fire off an alert to us. If we see little-to-no triggering by end users, we could drop them sooner rather than later.
+
+### Future Work
+
+The two-way mode plus transaction-like logic resolves the issue with the `code-` modules (or anything that triggers side-effect mutations in the cloud) while also still allowing the more diligent to review the implications of their changes. But we can do better.
+
+First, while entering the transaction mode introduces a synchronous point (such that `CALL iasql_commit();` blocks until the apply/sync operation is complete), there may be times where a user made changes in non-transactional two-way mode but they now need to wait for resources to actually be created. A proposed helper function would trigger a blocking call until a user condition is met. One possible syntax for this is:
+
+```sql
+SELECT iasql_wait_for('table_name', 5, 'cloud_id IS NOT NULL'); -- Default version
+
+SELECT iasql_wait_for('table_name', 5, 'cloud_id IS NOT NULL', 30); -- Optional argument for maximum # of seconds to wait before erroring (defaulting to 5 minutes or something)
+
+SELECT iasql_wait_for_all(); -- This uses special querying of the audit log to make sure all changes up to the time this function is invoked are represented in the cloud
+```
+
+The `iasql_wait_for` functions essentially execute (on another postgres thread/process) a particular select statement in a loop with a sleep thrown in until success or it errors out if the condition is never met. That select statement generated would be something like:
+
+```sql
+SELECT count(*) > 0 FROM `table_name` WHERE id = `5` AND `cloud_id IS NOT NULL`
+```
+
+The `iasql_wait_for_all`
```

Comment:
an orphan text here?

Reply:
oh, weird. I had written more than that. Let me see if I can recover it.